### DISCUSSION TITLE
Open NARS for Applications (ONA) NLU Chatbot

### DISCUSSION THREAD

**Discussion by StephenClarkML** [AGI-course/agi-course.github.io#3] [Ideas]
**RE: Course Project:
For software projects:
If application-centered: You might want to think about cases where there are benefits of having general-purpose AI over systems only programmed and trained for a single purpose.**

In my previous work with ONA I have created a robot that can use information from a vision system to move a robot to a goal using the ONA application with Narsese statements,  Reference Youtube video series. ( https://www.youtube.com/watch?v=uSywIEXRxmg&list=PLa-qAnPsdT6mSERDdk49u25tRw7UMw9P2 )

I am interested in trying to use ONA for Natural Language Understanding (NLU) in the form of a chatbot to add to my existing robot.

Up to this point I have only been able to get simple sentences into ONA using the python scripts built by Dr. Patrick Hammer.  English to Narsese ( https://github.com/opennars/OpenNARS-for-Applications/blob/master/english_to_narsese.py ) and the reverse ( https://github.com/opennars/OpenNARS-for-Applications/blob/master/narsese_to_english.py ).

For complex sentences I am looking into using this ML github repo ( https://github.com/iamarkaj/Split-and-Rephrase ) to split and rephrase long sentences into simple sentences.

I also want to see if the multi-step functionality built into ONA can be used in a NLU feedback loop so that the chatbot can reason about the statements sent to it.

I am also interested to find out how much of the built-in temporal and operations capabilities from the "goal based procedure learning " can be used to extend the NLU for the chatbot.

This would allow the "verbs" in a sentence that do not lemmatize to "is" to be converted to an action based on the verb name like ^run.

Example:
"Cat runs after a mouse" converts to something like:
< cat > &/ ^run =/> mouse

which roughly translates to:
Cat and run leads to mouse

I am still in the early stages of this idea so I welcome any comments.

Thanks 
Stephen

**Comment by patham9** [comment]
Exciting project! ML models (especially based on Transformers) could indeed be used to significantly improve over the rudimentary English-to-Narsese translation script english_to_narsese.py.
And adding procedural semantics / pragmatics to the picture can be very useful for robotics applications.
